### DISCUSSION TITLE
Possibility for query strings?

### DISCUSSION THREAD

**Discussion by 2000Nic** [zauberzeug/nicegui#870] [Q&A] [answered]
### Question

I know it's possible to create path parameters like [in the documentation](https://nicegui.io/documentation#pages_with_path_parameters), but is it possible to use them like GET method query strings?
For example: `127.0.0.1:8080/TLC?parameter=True` instead of `127.0.0.1:8080/TLC/True`
I'm asking as my page uses static files in a way that a path parameter set with a `/` wouldn't work because of some static files called in a header which is universal to all pages in the program.

**Comment by rodja** [comment, marked as answer]
You can do it the same way [as you would with FastAPI](https://fastapi.tiangolo.com/tutorial/query-params/?h=query#query-parameters):

```py
from nicegui import ui

@ui.page('/')
async def index(name: str = 'Test'):
    ui.button(name, on_click=lambda: ui.notify('Bam!'))

ui.run()
```

Open with https://localhost:8080/?name=Trigger.

**Reply by 2000Nic** [reply]
Thanks a lot @rodja!

**Comment by ed2050** [comment]
@rodja - There's a problem with nicegui's query param implementation.  Second instance of query param shadows previous value.  Python stdlib behavior is to return an array of values for each param.

# Example
```py
@ui.page ('/foo')  
def mypage (name = none) :  # get `name` from query params
    ui.label (f'{ name = }')

ui.run ()
# go to server/foo?name=foo : shows "name = foo"
# go to server/foo?name=foo&name=bar : shows "name = bar"
```

# Correct behavior

Python stdlib behavior is to create an array for each query param so multiple values are retained.  it would be better to follow stdlib practice.  I realize this may be beyond your control and may need to be fixed in fastapi.

```py
from urllib.parse import parse_qs

parse_qs ('name=foo')
# returns {'name': ['foo']}

parse_qs ('name=foo&name=bar')
# returns {'name': ['foo', 'bar']}
```

Temporary fix is to access raw url at `request.url` and apply `urlparse` & `parse_qs` from `urllib.parse`.

**Reply by falkoschindler** [reply]
Interesting observation. But NiceGUI isn't parsing the URL itself, but uses the underlying FastAPI framework. And the observed behavior seems to be the default in FastAPI. See https://fastapi.tiangolo.com/tutorial/query-params-str-validations/#query-parameter-list-multiple-values for more information.

**Comment by ed2050** [comment]
I expected so.  That's fine, but it should be fixed upstream.

FYI it's not just python stdlib.  The [URL DOM Standard](https://url.spec.whatwg.org/#dom-urlsearchparams-getall) specifies:
- get() to return the _first_ value of a named parameter
- getAll() to return multiple values

NG & FastAPI return the last value, which is not defined.
